### ISSUE TITLE
request, allow graphs to have settable background color

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1833]
right now, the only way to mimmick a background color is with an iemguis canvas object, either behind it or inside it. I'm used to having canvases behind arrays to give an impression of a colored array, but that involves some hassle. 

I wish there was a simpler way to set colors of 'GOP' subpatches, including arrays but this would also work for any GOP subpatch/abstraction.

How crazy is this?

hmmmm, now that I think of it, this also may imply that we get to set background color of any canvas, like even no GOP ones, which is also not a bad idea.

**Comment by porres** [comment]
@sebshader ?

**Comment by sebshader** [comment]
technically, I don't think it's that big of a thing but using iemgui canvasses for it is already a fine and flexible solution as well imo.

**Comment by porres** [comment]
> technically, I don't think it's that big of a thing but using iemgui canvasses for it is already a fine and flexible solution as well imo.

the problem is that it hides the inlets/outlets!

**Comment by porres** [comment]
also, if you resize the graph you need to resize the canvas and stuff...

**Comment by porres** [comment]
hey, here's an idea, if inlets/outlets are the main issue, why not foce them to always draw on top? Is it possible? what do you say @sebshader ?

**Comment by sebshader** [comment]
The pd GOP draws things on the parent Canvas so I think you would have to draw a rectangle behind the GOP object somehow

**Comment by porres** [comment]
> The pd GOP draws things on the parent Canvas so I think you would have to draw a rectangle behind the GOP object somehow

so this means that no 'gui message' could do the trick, right? You'd have to deliberately create a rectangle and put it in the background. With the message, the background color of the GOP is the same as the parent patch's

![Screen Shot 2023-01-30 at 03 02 51](https://user-images.githubusercontent.com/5360142/215598872-449270d2-a68a-4056-9c67-79947e3e040d.png)

**Comment by sebshader** [comment]
right, the canvas configure stuff affects the whole canvas. and in pd everything in a patch window is drawn on the same canvas widget, even if those things are GOP.

**Comment by porres** [comment]
And what about forcing inlets to be drawn always on top of anything in a 'GOP'? How hard would it be @sebshader? And do you think people who used canvases that hid inlets/outlets would potentially  mind the change? I don't see it as anything that would actually 'break' anything, what do you say @umlaeute?

**Comment by sebshader** [comment]
personally I like the idea of drawing inlets on top for GOP regardless of color vs no color tbh
Maybe it's as easy as moving the ```glist_drawiofor``` call to the end of ```graph_vis``` (in g_graph.c)

**Comment by ulioidle** [comment]
> And do you think people who used canvases that hid inlets/outlets would potentially mind the change? I don't see it as anything that would actually 'break' anything,

I have some abstractions which this would be irksome on but it would be low on my list of gripes about pd and could live with it as long as no color actually means no color and they will be transparent. Perhaps a toggle in the GOP dialog to hide i/o?

**Comment by umlaeute** [comment]
not sure about the usefulness of hiding iolets (esp. as a property of a the object itself (as opposed to a property of the patch the object lives in)

**Comment by ulioidle** [comment]
My thinking was that it is fairly common to hide the gop box with a canvas, if the box is hidden this way but iolets are always drawn on top that could result in what looks like floating iolets if the canvas is the same color as the patch background, this could get confusing in some situations. I have a few abstraction which this would happen in but as I said a minor issue for me and simple enough to redesign those few abstractions.

**Comment by umlaeute** [comment]
i don't think that would be a good idea.

i would not expect the "always on top" iolets to be on top of *everything*, so your use-case is a good one to check when "this" is implemented.

**Comment by porres** [comment]
>  was that it is fairly common to hide the gop box with a canvas

what do you mean?

**Comment by ulioidle** [comment]
@porres Hiding some or all of frame with a canvas in the abstraction. 
![abs](https://github.com/pure-data/pure-data/assets/8304564/15299ef0-6dad-4c84-981c-9eb3cf53b2d4)

**Comment by porres** [comment]
what's the use case of that? :)

**Comment by ulioidle** [comment]
Partially just aesthetics and a reflection of personal style but it also allows for making cleaner interfaces with less visual clutter that reflect the needs of the interface/performance instead of the source level/patching. Everything in a box is useful for patching but less useful on the performance side, it can be difficult to quickly parse a screen filled with boxes full of boxes. I have one patch which moves the canvas out of the GOP area for the object which my midi controller is currently editing so only the abstraction which is being worked on is in a box, a very simple and effective solution which retains the easy to read high contrast of back on white and plays well with those elements whose color we can not set.

**Comment by porres** [comment]
I see that as quite an edge case and some not fairly common at all, never ever seen it, looks like a quite personal use and I hope you don't mind adapting and finding new tricks...

Anyway, I'm a bit lost now. How hard is it to just enable a background color for GOPs? That was the original request anyway. Doing this without hiding inlets/outlets of course

**Comment by porres** [comment]
> How hard is it to just enable a background color for GOPs? That was the original request anyway. Doing this without hiding inlets/outlets of course

for the record @sebshader tells me on the discord channel this is quite possible. I vote for this as originally requested cause I think [cnv] is kinda lame in general and one major convenience would be to simply change the area size of a GOP (via properties or via clicking and dragging in edit mode) and the background simply adjusts.
